larray/inout/pandas.py

```diff
@@ -39,34 +39,39 @@ def parse(s):
         return s
 
 
-def df_labels(df, sort=True):
+def index_to_labels(idx, sort=True):
     """
     Returns unique labels for each dimension.
     """
-    idx = df.index
     if isinstance(idx, pd.core.index.MultiIndex):
         if sort:
             return list(idx.levels)
         else:
             return [list(unique(idx.get_level_values(l))) for l in idx.names]
     else:
         assert isinstance(idx, pd.core.index.Index)
-        # use .values if needed
-        return [idx]
+        return [list(idx.values)]
 
 
 def cartesian_product_df(df, sort_rows=False, sort_columns=False, **kwargs):
-    labels = df_labels(df, sort=sort_rows)
-    if sort_rows:
-        new_index = pd.MultiIndex.from_product(labels)
+    idx = df.index
+    labels = index_to_labels(idx, sort=sort_rows)
+    # since version 0.21 of pandas, the MultiIndex constructors no longer squeezes
+    # a MultiIndex with all length-one levels down to a regular Index.
+    if isinstance(idx, pd.core.index.MultiIndex):
+        if sort_rows:
+            new_index = pd.MultiIndex.from_product(labels)
+        else:
+            new_index = pd.MultiIndex.from_tuples(list(product(*labels)))
     else:
-        new_index = pd.MultiIndex.from_tuples(list(product(*labels)))
+        new_index = idx
```

Comment:
I guess `sort_rows=True` is not respected in this case.

Reply:
All tests are all green. This means, I guess, that a new test with `sort_rows` must be added to cover this case.